File: JS/ReactionDiffusionAlg/Diffusion.js
// Source:  http://karlsims.com/rd.html

class Diffusion {
  constructor() {
    this.width = Math.floor(W / step);
    this.height = Math.floor(H / step);

    // Some typical values
    this.diffusion = { a: 1.0, b: 0.5 };
    this.feed = 0.055;
    this.kill = 0.062;
    this.weight = [
      [0.05, 0.2, 0.05],
      [0.2, -1, 0.2],
      [0.05, 0.2, 0.05],
    ];

    this.curr = [];
    this.next = [];

    for (let i = 0; i < this.height; i++) {
      this.curr.push([]);
      this.next.push([]);
      for (let j = 0; j < this.width; j++) {
        this.curr[i][j] = { a: 1, b: 0 };
        this.next[i][j] = { a: 1, b: 0 };
      }
    }

    let x = Math.floor(this.width / 2) - 5;
    let y = Math.floor(this.height / 2) - 5;
    this.addChemical(x, y, 10);
  }

  addChemical(x, y, range = 10) {
    for (let i = 0; i < range; i++) for (let j = 0; j < range; j++) this.curr[y + i][x + j].b = 1;
  }

  update(i, j) {
    if (i == 0 || j == 0 || i == this.width - 1 || j == this.height - 1) return;

    let { a, b } = this.curr[i][j];

    let lap = this.LaplacianFunc(int(j), int(i));

    this.next[i][j].a = a + (this.diffusion.a * lap.a - a * b * b + this.feed * (1 - a));
    this.next[i][j].b = b + (this.diffusion.b * lap.b + a * b * b - (this.kill + this.feed) * b);
  }

  LaplacianFunc(x, y) {
    let a = 0;
    let b = 0;

    for (let i = -1; i < 2; i++) {
      for (let j = -1; j < 2; j++) {
        a += this.curr[i + y][j + x].a * this.weight[i + 1][j + 1];
        b += this.curr[i + y][j + x].b * this.weight[i + 1][j + 1];
      }
    }
    return { a, b };
  }

  show() {
    noStroke();

    for (let i in this.curr)
      for (let j in this.curr) {
        this.update(i, j);

        let { a, b } = this.next[i][j];
        let color = Math.floor((a - b) * 255);

        fill(color);
        rect(j * step, i * step, step, step);
      }

    this.swap();
  }

  swap() {
    let temp = this.curr;
    this.curr = this.next;
    this.next = temp;
  }
}

File: python/ReactionDiffusionAlg/Diffusion.py
from PIL import Image, ImageDraw


class Diffusion:
    curr = []
    next = []

    height = 200
    width = 200

    epoch = 0

    diffusion = {'A': 1, 'B': 0.5}
    feed = 0.0749
    kill = 0.0625

    weight = [
        [0.05, 0.2, 0.05],
        [0.2,  -1,   0.2],
        [0.05, 0.2, 0.05],
    ]

    def __init__(self):

        for i in range(self.height):
            self.curr.append([])
            self.next.append([])

            for j in range(self.width):
                self.curr[i].append({'A': 1, 'B': 0})
                self.next[i].append({'A': 1, 'B': 0})

        self.addChemical(self.width // 2, self.height // 2, 10)

    def addChemical(self, x, y, range_=10):
        for i in range(-range_, range_):
            for j in range(-range_, range_):
                self.curr[y + i][x + j]['B'] = 1

    def update(self):
        for i in range(1, self.height - 1):
            for j in range(1, self.width - 1):

                a = self.curr[i][j]['A']
                b = self.curr[i][j]['B']

                laplace = self.LaplacianFunc(j, i)

                self.next[i][j]['A'] = a + \
                    (self.diffusion['A'] * laplace['A'] -
                     a * b * b + self.feed * (1 - a))
                self.next[i][j]['B'] = b + \
                    (self.diffusion['B'] * laplace['B'] + a * b *
                     b - (self.kill + self.feed) * b)

        # Swap
        self.next, self.curr = self.curr, self.next
        self.epoch += 1

    def LaplacianFunc(self, x, y):
        result = {'A': 0, 'B': 0}

        for i in range(-1, 2):
            for j in range(-1, 2):
                result['A'] += self.curr[i + y][j + x]['A'] * \
                    self.weight[i + 1][j + 1]
                result['B'] += self.curr[i + y][j + x]['B'] * \
                    self.weight[i + 1][j + 1]

        return result

    def save(self, name='Diffusion'):
        img = Image.new('RGB', (self.width, self.height))

        pixels = []

        for i in range(self.height):
            for j in range(self.width):
                a = self.curr[i][j]['A']
                b = self.curr[i][j]['B']
                color = round((a - b) * 255)

                pixels.append((color, color, color))
        img.putdata(pixels)
        img.save(name + '.png')
